core/src/main/java/org/apache/accumulo/core/data/KeyBuilder.java

```diff
@@ -0,0 +1,300 @@
+     *
+     * @param columnFamily
+     *          the column family to use for the {@link Key}
+     * @return this builder
+     */
+    ColumnQualifierStep family(final byte[] columnFamily);
+
+    /**
+     * Set the column family of the {@link Key} that this builder will build to the parameter.
+     *
+     * @param columnFamily
+     *          the column family to use for the {@link Key}
+     * @return this builder
+     */
+    ColumnQualifierStep family(final Text columnFamily);
+
+    /**
+     * Set the column family of the {@link Key} that this builder will build to the parameter.
+     *
+     * @param columnFamily
+     *          the column family to use for the {@link Key}
+     * @return this builder
+     */
+    ColumnQualifierStep family(final CharSequence columnFamily);
+  }
+
+  public interface ColumnQualifierStep extends ColumnVisibilityStep {
+
+    /**
+     * Set the column qualifier of the {@link Key} that this builder will build to the parameter.
+     *
+     * @param columnQualifier
+     *          the column qualifier to use for the {@link Key}
+     * @return this builder
+     */
+    ColumnVisibilityStep qualifier(final byte[] columnQualifier);
+
+    /**
+     * Set the column qualifier of the {@link Key} that this builder will build to the parameter.
+     *
+     * @param columnQualifier
+     *          the column qualifier to use for the {@link Key}
+     * @return this builder
+     */
+    ColumnVisibilityStep qualifier(final Text columnQualifier);
+
+    /**
+     * Set the column qualifier of the {@link Key} that this builder will build to the parameter.
+     *
+     * @param columnQualifier
+     *          the column qualifier to use for the {@link Key}
+     * @return this builder
+     */
+    ColumnVisibilityStep qualifier(final CharSequence columnQualifier);
+  }
+
+  public interface ColumnVisibilityStep extends Build {
+
+    /**
+     * Set the column qualifier of the {@link Key} that this builder will build to the parameter.
+     *
+     * @param columnVisibility
+     *          the column visibility to use for the {@link Key}
+     * @return this builder
+     */
+    Build visibility(final byte[] columnVisibility);
+
+    /**
+     * Set the column qualifier of the {@link Key} that this builder will build to the parameter.
+     *
+     * @param columnVisibility
+     *          the column visibility to use for the {@link Key}
+     * @return this builder
+     */
+    Build visibility(final Text columnVisibility);
+
+    /**
+     * Set the column qualifier of the {@link Key} that this builder will build to the parameter.
+     *
+     * @param columnVisibility
+     *          the column visibility to use for the {@link Key}
+     * @return this builder
+     */
+    Build visibility(final CharSequence columnVisibility);
+  }
+
+  static class AbstractKeyBuilder implements RowStep, ColumnFamilyStep, ColumnQualifierStep,
```

Comment:
`KeyBuilder` is the name of the enclosing class, could call it `KeyBuilderImpl`.